stdlibs/stdlibs.go

```diff
@@ -141,6 +141,61 @@ func InjectPackage(store gno.Store, pn *gno.PackageNode) {
 				m.PushValue(res0)
 			},
 		)
+		// TODO: IBC handshake process
+		// this is called mannuly, maybe should have some modifier for the contract,like `IBC`,
+		// so the IBC channel is initialized during deploying, and can be used more efficiently
+		pn.DefineNative("OpenIBC",
+			gno.Flds( // params
+				"", "string",
+			),
+			gno.Flds( // results
+				"", "string",
+			),
+			func(m *gno.Machine) {
+				// TODO: the handshake logic
+			},
+		)
+		pn.DefineNative("Call",
+			gno.Flds( // params
+				"pkg", "string",
+				"fn", "string",
+				"arg", "string",
+			),
+			gno.Flds( // results
+				"r", "string",
+			),
+			func(m *gno.Machine) {
+				// get args
+				arg0, arg1, arg2 := m.LastBlock().GetParams3()
+				pkgPath := arg0.TV.GetString()
+				fnc := arg1.TV.GetString()
+				arg := arg2.TV.GetString()
+
+				// construct msg
+				msg := &CallMsg{
+					PkgPath: pkgPath,
+					Fn:      fnc,
+					Args:    []string{arg},
+				}
+				// res0 := gno.Go2GnoValue(
+				// 	m.Alloc,
+				// 	m.Store,
+				// 	reflect.ValueOf(msg),
+				// )
+				// m.PushValue(res0)
+
+				go SendData(msg)
```

Comment:
We have two ways to achieve this, like using the channel to send and receive msg, with timeout set properly,  this is  more like  a synchronous call, it will block until  the call is finished.

Another ways using a callback, that the VM keeper initiate a callback to the caller after the target call is finished, this is more like an asynchronous call, but brings complexity compared to the first one.

Both have pros and cons,  what do you think.

Reply:
I'm considering eliminating the goroutine since we could have events in a queue without waiting.

Can you begin implementation as per your preference, and we can discuss it later? Perhaps I can propose an alternative solution.

Reply:
Here is a prototype: 

**Contract:**

The `caller` contract is something like this, it firstly encodes the parameters, invoke a std.Await(exactly name to be determined, name it `await` since it's synchronous call"), and wait for the result.

```
 package hello

 import "std"

 func Hello() string {

 	chainID := std.GetChainID()
 	call := std.EncodeMsgCall(chainID, "gno.land/r/demo/x/calls/await/greet", "Greet", []string{"hey"})

 	r, err := std.Await(call)
 	println("done ")
 	println("r is :", r)
 	println("errMsg is: ", err)

 	return "hello" + r
}
```

The encoder is very rough, just a simple stringification like:

```
package std

 // TODO: a better encoder
 func EncodeMsgCall(ChainID, PkgPath, Fn string, args []string) string {
 	var as string
 	for _, a := range args {
 		as += a
 	}
 	return ChainID + "#" + PkgPath + "#" + Fn + "#" + as
 }

```

Note: the `ChainID` is used temperarily to identify whether  this is a in-VM call or IBC call.



**Stdlib:**

The stringified `call` is decoded into a `GnoMsg`, which mostly resembles `MsgCall` but with additionally  field of `ChainID` , and   field `Response`, that is a `chan string` as a callback to retrieve result.

```
pn.DefineNative("Await",
 			gno.Flds( // params
 				"call", "string",
 			),
 			gno.Flds( // results
 				"bz", "string",
 				"err", "string",
 			),
 			func(m *gno.Machine) {
 				// println("std.Send")
 				arg0 := m.LastBlock().GetParams1()
 				call := arg0.TV.GetString()
 				// println("call: ", call)

 				gnoMsg, err := vmh.DecodeMsg(call)
 				if err != nil {
 					panic("parameter invalid")
 				}
 				resQueue := make(chan string, 1)
 				gnoMsg.Response = resQueue

 				// send msg
 				vmkeeper.DispatchInternalMsg(gnoMsg)

 				// XXX: how this determined, since calls will accumulate
 				// should have an estimation like gas estimation?
 				timeout := 3 * time.Second
 				println("block, waiting for result...")

 				// TODO: err handling
 				var result string
 				select {
 				case result = <-resQueue:
 					println("callback in recvMsg: ", result)
 				case <-time.After(timeout):
 					panic("time out")
 					// case err = <- errQueue:
 				}

 				// TODO: return err to contract
 				var errMsg string
 				if err != nil {
 					errMsg = err.Error()
 				}

 				res := gno.Go2GnoValue(
 					m.Alloc,
 					m.Store,
 					reflect.ValueOf(result),
 				)
 				m.PushValue(res)
 				m.PushValue(typedString(gno.StringValue(errMsg)))
 			},
 		)
```

**VMKeeper:**

A eventLoop routine is always listening new event from contract, and then handle it.

```
func (vmk *VMKeeper) startEventLoop() {
	for {
		select {
		case msg := <-vmk.internalMsgQueue:
			go vmk.HandleMsg(msg)
		}
	}
}
```

The handler calls the `callee` contract in-VM or through IBC. We don't have IBC now, so using a  channel to simulate the IBC loop.

```
func (vmk *VMKeeper) HandleMsg(msg vmh.GnoMsg) {
	println("------HandleMsg, routine spawned ------")
	// prepare call
	msgCall, isLocal, response, err := vmk.preprocessMessage(msg)
	if err != nil {
		panic(err.Error())
	}
	// do the call
	if isLocal {
		println("in VM call")
		println("msgCall: ", msgCall.Caller.String(), msgCall.PkgPath, msgCall.Func, msgCall.Args[0])

		r, err := vmk.Call(vmk.ctx, msgCall)
		println("call finished, res: ", r)
		// have an return
		if err == nil {
			response <- r
		}
	} else { // IBC call
		println("IBC call")
		// send IBC packet, waiting for OnRecv
		vmk.ibcResponseQueue = response
		vmk.SendIBCMsg(vmk.ctx, msgCall)
	}
}
```

So it's a basic prototype, is it the right path?

**To be determined and done:**

1. Sync or Async, if it's a cascade call, like a->b->c, the return value would be useful so a synchronous call will be needed, but if it's something like logs, events (with no intended action next), a Async call will be ok.
2. The proper name ;
3. the `std.GetXXXCaller` related work. this may relate to some security issues, should get whole call graph to solve the impersonate issues?
4. gas management;
5. more test.